Tag Symata v0.0.17 [https://github.com/jlapeyre/Symata.jl]

jlapeyre opened the pull request:
* fixed typesquatting regression `sizeof(::Symbol)`
* initial support for Jupyter
* initial support for `Plots.jl`

tkelman commented:
Note that doing conditional dependencies like you're doing here doesn't interact well with precompilation.

jlapeyre commented:
I'll have to look into what the problems are. With other, unrelated, modules I sometimes get an error due to stale, cached compiled code and have to restart Juila. Can it get worse than this ?

tkelman commented:
If the set of packages available at precompile time does not match the set of packages available at load time, you will either get missing functionality or errors. You can fix this by either disabling precompilation, making dependencies unconditional, or moving optional dependencies to separate packages.

jlapeyre commented:
I don't see either of these problems arising. I can do the following.

* Install `Plots`, `PyPlot`, `IJulia`. 
* Delete cached compilation of `Symata`.
* do `using Symata` and call a function that uses `Plots`. At this point `Plots` is precompiled and I get the expected result.
* Remove `Plots`,  `PyPlot`, `GR`, `IJulia`, all packages with the word  "plot" in the title.
* Delete the cached precompilation of each of these removed packages
* Check the package directory to verify that none of these packages is installed.
* Run `Symata` and try a plot. I get the message asking me to install `Plots`, which is what I expect.
* At this point, I can install the plotting packages and they work again with `Symata`. Or, I can delete
the cached `Symata.ji`, then run `Symata` and see that plotting works.

I can also do this.

* Remove all plotting packages and associated `.jl` files.
* Verify that these package directories are gone.
* Delete `Symata.ji`. Run `Symata` to regenerate `Symata.ji`.
* Install `Plots.jl`.
* Run `Symata` and ask for a plot. The relevant packages are precompiled and the plot is made.

I see the same story with `IJulia`.

Does the following make sense ? : Suppose package A checks while being loaded (and precompiled) for the presence of another package B and does `using B`  if it is present (or uses try catch). If B was installed when A was precompiled and you then remove B, then you get an error the next time you use A. Or, if you precompile package A when B is not installed, and later install B, the information on the absence of B is (in a sense) already in `A.ji`, so B is not loaded.

It never occurred to me to try something like this because the problem seems obvious. (Obvious now. I vaguely recall learning something similar  in the past, probably with Common Lisp.) 

 `Symata` does not check for the presence of conditional dependencies when it is being precompiled. It checks for the dependency and uses `using` only while running, the first time a function is called that uses the dependency.

Of course, there could be some issue that I am overlooking here.

jlapeyre commented:
But, there are a couple of other problems with the PR that are fixed in the latest [commit](https://github.com/jlapeyre/Symata.jl/commit/df95edc59a760c3f4710179d72dbd1179a2acd24) to master in `Symata`. 

* I inadvertently made the interface with `IJulia.jl`  for the master branch of `IJulia.jl`, not the last tagged release. I have fixed this.

* I left a couple of diagnostic print statements in the code. I have removed these.
